### ISSUE TITLE
redundant_clone: suggestion breaks code if cloned value would be dropped immediately (not bound to a var)

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#7429]
Lint name: redundant_clone

Inside rust analyzer I found code like this:

```rust

pub fn bad() -> String {
    let s = String::new();
    s.to_string();
    s
}
```
````
warning: redundant clone
 --> src/main.rs:7:6
  |
7 |     s.to_string();
  |      ^^^^^^^^^^^^ help: remove this
  |
  = note: `#[warn(clippy::redundant_clone)]` on by default
note: cloned value is neither consumed nor mutated
 --> src/main.rs:7:5
  |
7 |     s.to_string();
  |     ^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
  ````

Clippy correctly detects the redundant clone and suggests removing it, but it only suggests removing the `clone()`ing.
The problem is we do not bind the cloned value to anything here and replacing `s.to_string();` with `s;` drops `s` and it can no longer be returned. :laughing: 

````rust
pub fn bad() -> String {
    let s = String::new();
    s;
    s
}
````

In such a case (clone that is neither bound to a variable nor returned) I think we can remove the entire statement, can't we?

### Meta
`clippy 0.1.55 (798baeb 2021-07-02)`

**Comment by lengyijun** [comment]
A quick fix:  change redundant_clone.rs:427 to
```
 clone_consumed_or_mutated: true
```
But this makes tests failed.
